### PR TITLE
Chapter 3 created

### PR DESCRIPTION
**Pull request by agustinaGallo97**
## :link: Jira Tracker reference:
*[AP - Chapter 3] The Activity Lifecycle - https://androidtraining.atlassian.net/browse/AT-40

## Description:
I've done chapter 3 and also both challenges.

## :camera_flash: UI preview
ezgif com-video-to-gif(3)

## :pushpin: Notes:
Both challenges are included.

### CODE REVIEW

File: GeoQuizChallenge/app/src/main/AndroidManifest.xml [hunk 1]
```diff
@@ -13,7 +13,7 @@
         android:theme="@style/AppTheme"
         >
 
-        <activity android:name=".QuizActivity">
+        <activity android:name=".Views.QuizActivity">
```
Comment (mirland) [inline GeoQuizChallenge/app/src/main/AndroidManifest.xml]: use lowercase for packages

File: GeoQuizChallenge/app/src/main/java/com/bignerdranch/android/geoquiz/Controllers/QuestionController.java [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+import com.bignerdranch.android.geoquiz.R;
+import java.util.ArrayList;
+import java.util.List;
+
+public class QuestionController {
+  private static final QuestionController INSTANCE = new QuestionController();
+  private List<Question> questionBank;
+
+  private QuestionController() {
+    questionBank = new ArrayList<Question>();
```
Comment (mirland) [inline GeoQuizChallenge/app/src/main/java/com/bignerdranch/android/geoquiz/Controllers/QuestionController.java]: you can use `Arrays.asList(`

File: GeoQuizChallenge/app/src/main/java/com/bignerdranch/android/geoquiz/Controllers/QuestionController.java [hunk 3]
```diff
@@ -0,0 +1,29 @@
+package com.bignerdranch.android.geoquiz.Controllers;
+
+import com.bignerdranch.android.geoquiz.Models.Question;
+import com.bignerdranch.android.geoquiz.R;
+import java.util.ArrayList;
+import java.util.List;
+
+public class QuestionController {
+  private static final QuestionController INSTANCE = new QuestionController();
+  private List<Question> questionBank;
```
Comment (mirland) [inline GeoQuizChallenge/app/src/main/java/com/bignerdranch/android/geoquiz/Controllers/QuestionController.java]: this should be `final` and `@NonNull`

File: GeoQuizChallenge/app/src/main/java/com/bignerdranch/android/geoquiz/Controllers/QuestionController.java [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+    questionBank = new ArrayList<Question>();
+    questionBank.add(new Question(R.string.question_australia, true));
+    questionBank.add(new Question(R.string.question_oceans, true));
+    questionBank.add(new Question(R.string.question_mideast, true));
+    questionBank.add(new Question(R.string.question_africa, true));
+    questionBank.add(new Question(R.string.question_americas, true));
+    questionBank.add(new Question(R.string.question_asia, true));
+  }
+
+  public static QuestionController getInstance() {
```
Comment (mirland) [inline GeoQuizChallenge/app/src/main/java/com/bignerdranch/android/geoquiz/Controllers/QuestionController.java]: Nullability

File: GeoQuizChallenge/app/src/main/java/com/bignerdranch/android/geoquiz/Controllers/QuestionController.java [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+    return INSTANCE;
+  }
+
+  public List<Question> getQuestionBank() {
```
Comment (mirland) [inline GeoQuizChallenge/app/src/main/java/com/bignerdranch/android/geoquiz/Controllers/QuestionController.java]: Nullability

File: GeoQuizChallenge/app/src/main/java/com/bignerdranch/android/geoquiz/Views/QuizActivity.java [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+  @SuppressWarnings("FieldCanBeLocal")
+  private Button falseButton;
+  private ImageButton nextButton;
+  private ImageButton prevButton;
+  private TextView questionTextView;
+  private List<Question> questionBank = QuestionController.getInstance().getQuestionBank();
+  private int currentIndex = 0;
+  private int countOfCorrectQuestionsResolved = 0;
+  private int countOfQuestionsResolved = 0;
+  private HashSet<Question> resolvedQuestions = new HashSet();
```
Comment (mirland) [inline GeoQuizChallenge/app/src/main/java/com/bignerdranch/android/geoquiz/Views/QuizActivity.java]: I think you can define it as set
```suggestion
  private Set<Question> resolvedQuestions = new HashSet();
```

File: GeoQuizChallenge/app/src/main/java/com/bignerdranch/android/geoquiz/Views/QuizActivity.java [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+    if (resolvedQuestions.contains(actualQuestion)) {
+      disableButtons(false);
+    } else {
+      @StringRes
+      int answerTest = isCorrect ? R.string.correct : R.string.incorrect;
+      Toast toast = Toast.makeText(this, answerTest, Toast.LENGTH_SHORT);
+      toast.setGravity(Gravity.TOP | Gravity.CENTER_HORIZONTAL, 0, 0);
+      toast.show();
+      if (isCorrect) {
+        countOfCorrectQuestionsResolved += 1;
```
Comment (mirland) [inline GeoQuizChallenge/app/src/main/java/com/bignerdranch/android/geoquiz/Views/QuizActivity.java]: the method name is `showToastCorrectly` but you are doing a lot of thing. Please, split it in multiple methods

File: GeoQuizChallenge/app/src/main/java/com/bignerdranch/android/geoquiz/Views/QuizActivity.java [hunk 8]
```diff
@@ -0,0 +1,154 @@
...
+      }
+      countOfQuestionsResolved += 1;
+      resolvedQuestions.add(actualQuestion);
+      disableButtons(false);
+    }
+  }
+
+  private void disableButtons(boolean value) {
```
Comment (mirland) [inline GeoQuizChallenge/app/src/main/java/com/bignerdranch/android/geoquiz/Views/QuizActivity.java]: the param name is not good (it should be representative)

Comment (mirland) [reply GeoQuizChallenge/app/src/main/java/com/bignerdranch/android/geoquiz/Views/QuizActivity.java]: Also we try to add positive names in methods, maybe you can change `disableButtons` to `enableButtons`. However, I think `setButtonsState` is better

File: GeoQuizChallenge/app/src/main/java/com/bignerdranch/android/geoquiz/controllers/QuestionController.java [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+import com.bignerdranch.android.geoquiz.R;
+import java.util.Arrays;
+import java.util.List;
+
+import androidx.annotation.NonNull;
+
+public class QuestionController {
+  @NonNull
+  private static final QuestionController INSTANCE = new QuestionController();
+  private List<Question> questionBank;
```
Comment (mirland) [inline GeoQuizChallenge/app/src/main/java/com/bignerdranch/android/geoquiz/controllers/QuestionController.java:13]: Nullability

File: GeoQuizChallenge/app/src/main/java/com/bignerdranch/android/geoquiz/views/QuizActivity.java [hunk 10]
```diff
@@ -34,7 +35,7 @@
   private int currentIndex = 0;
   private int countOfCorrectQuestionsResolved = 0;
   private int countOfQuestionsResolved = 0;
-  private HashSet<Question> resolvedQuestions = new HashSet();
+  private Set<Question> resolvedQuestions = new HashSet();
```
Comment (mirland) [inline GeoQuizChallenge/app/src/main/java/com/bignerdranch/android/geoquiz/views/QuizActivity.java]: nullability and final

File: GeoQuizChallenge/app/src/main/java/com/bignerdranch/android/geoquiz/views/QuizActivity.java [hunk 11]
```diff
@@ -116,9 +117,9 @@ private void updateQuestion() {
     questionTextView.setText(actualQuestion.getTextResId());
 
     if (resolvedQuestions.contains(actualQuestion)) {
-      disableButtons(false);
+      setButtonsState(false);
     } else {
-      disableButtons(true);
+      setButtonsState(true);
     }
```
Comment (mirland) [inline GeoQuizChallenge/app/src/main/java/com/bignerdranch/android/geoquiz/views/QuizActivity.java]: you can make it shorter

File: GeoQuizChallenge/app/src/main/java/com/bignerdranch/android/geoquiz/views/QuizActivity.java [hunk 12]
```diff
@@ -127,26 +128,38 @@ private void updateQuestion() {
...
+    countOfQuestionsResolved += 1;
+  }
+
+  private void updateResolvedQuestions(Question actualQuestion) {
+    resolvedQuestions.add(actualQuestion);
+    setButtonsState(false);
   }
 
-  private void disableButtons(boolean value) {
+  private void setButtonsState(boolean value) {
```
Comment (mirland) [inline GeoQuizChallenge/app/src/main/java/com/bignerdranch/android/geoquiz/views/QuizActivity.java]: change the param name, maybe `enabled`?

**Review by mirland — APPROVED** [review]

File: GeoQuizChallenge/app/src/main/java/com/bignerdranch/android/geoquiz/views/QuizActivity.java [hunk 13]
```diff
@@ -0,0 +1,167 @@
...
+  }
+
+  private void increaseCounters(boolean isCorrect) {
+    if (isCorrect) {
+      countOfCorrectQuestionsResolved += 1;
+    }
+    countOfQuestionsResolved += 1;
+  }
+
+  private void updateResolvedQuestions(Question actualQuestion) {
```
Comment (FerrariBruno) [inline GeoQuizChallenge/app/src/main/java/com/bignerdranch/android/geoquiz/views/QuizActivity.java:159]: nullability

**Review by FerrariBruno — APPROVED** [review]

**Review by FerrariBruno — CHANGES_REQUESTED** [review]
It's missing Timber! Look at the card's description: https://androidtraining.atlassian.net/browse/AT-40

**Review by FerrariBruno — APPROVED** [review]

**Pull request merged**